### PR TITLE
Add ZIO#doUntil, doWhile, retryUntil and retryWhile equals and effectful variants

### PR DESCRIPTION
**Pull request by saraiva132**
A couple of combinators missing in ZIO that were present in scheduler that could be added for free.

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -1340,15 +1364,39 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
+   * Retries this effect until its error satisfies the specified effectful predicate.
+   */
+  final def retryUntilM(f: E => UIO[Boolean])(implicit ev: CanFail[E]): ZIO[R, E, A] =
+    retry(Schedule.doUntilM(f))
+
   /**
    * Retries this effect while its error satisfies the specified predicate.
    */
-  final def retryWhile(f: E => Boolean): ZIO[R, E, A] =
+  final def retryWhile(f: E => Boolean)(implicit ev: CanFail[E]): ZIO[R, E, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala]: Good catch!

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -1340,15 +1364,39 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
     retry(Schedule.doWhile(f))
 
+  /**
+   * Repeats this effect for as long as the error equals the predicate.
+   */
+  final def retryWhileEquals[E1 >: E](e: => E1)(implicit ev: CanFail[E1]): ZIO[R, E1, A] =
+    retry(Schedule.doWhileEquals(e))
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala]: You need to change the parameter to `Schedule.doWhileEquals` to be by name if you want this to actually be lazy.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 3]
```diff
@@ -1340,15 +1364,39 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
    */
-  final def retryUntil(f: E => Boolean): ZIO[R, E, A] =
+  final def retryUntil(f: E => Boolean)(implicit ev: CanFail[E]): ZIO[R, E, A] =
     retry(Schedule.doUntil(f))
 
+  /**
+   * Retries this effect until its error equals the predicate.
+   */
+  final def retryUntilEquals[E1 >: E](e: => E1)(implicit ev: CanFail[E1]): ZIO[R, E1, A] =
+    retry(Schedule.doUntilEquals(e))
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala]: Same comment as below regarding `Schedule.doUntilEquals`.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 4]
```diff
@@ -1340,15 +1364,39 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   /**
    * Retries this effect until its error satisfies the specified predicate.
    */
-  final def retryUntil(f: E => Boolean): ZIO[R, E, A] =
+  final def retryUntil(f: E => Boolean)(implicit ev: CanFail[E]): ZIO[R, E, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala]: Nice!

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 5]
```diff
@@ -377,12 +377,36 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   final def doUntil(f: A => Boolean): ZIO[R, E, A] =
     repeat(Schedule.doUntil(f))
 
+  /**
+   * Repeats this effect until its result is equal to the predicate.
+   */
+  final def doUntilEquals[A1 >: A](a: => A1): ZIO[R, E, A1] =
```
Comment (mijicd) [inline core/shared/src/main/scala/zio/ZIO.scala]: I believe you can omit the type bound here. Also, as Adam pointed out, `doUntilEquals` accepts parameter by value.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 6]
```diff
@@ -377,12 +377,36 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
   /**
    * Repeats this effect while its result satisfies the specified predicate.
    */
   final def doWhile(f: A => Boolean): ZIO[R, E, A] =
     repeat(Schedule.doWhile(f))
 
+  /**
+   * Repeats this effect for as long as the predicate is equal to its result.
+   */
+  final def doWhileEquals[A1 >: A](a: => A1): ZIO[R, E, A1] =
```
Comment (mijicd) [inline core/shared/src/main/scala/zio/ZIO.scala]: Same here.

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 5 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/ZIO.scala]: Ah, scratch it, forgot about the variance. Laziness it is :).

**Pull request merged**